modules/packages/Sort.chpl

```diff
@@ -18,27 +18,178 @@
+returns a value that supports all of comparison operations: ``<``, ``<=``,
+``>``, ``>=``, or ``==``.
+
+The default key method would look like this:
+
+.. code-block:: chapel
+
+  proc default.key(a) {
+    return a;
+  }
+
+
+As an example, if we want to sort an array by the absolute values of its
+elements, we can use the following comparator record with a key method:
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record comparator { }
+
+  // key method maps an element to the value to be used for comparison
+  proc comparator.key(a) { return abs(a); }
+
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+
+The ``compare(a, b)`` method accepts 2 arguments of the data ``eltType`` and
+returns an integer between ``1`` if ``a > b``, ``0`` if ``a == b``, and ``-1`` if ``a < b``.
+
+The default compare method would look like this:
+
+.. code-block:: chapel
+
+    proc default.compare(a, b) {
+      if a > b {
+        return 1;
+      } else if a < b {
+        return -1;
+      } else {
+        return 0;
+      }
+    }
+
+
+The absolute value comparison example from above can alternatively be
+implemented with a compare method:
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc comparator.compare(a, b) {
+    if abs(a) > abs(b) {
+      return 1;
+    } else if abs(a) < abs(b) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+If both methods are implemented on the record passed as the comparator, the
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
 
  */
+
 module Sort {
 
-private inline proc chpl_sort_cmp(a, b, param reverse=false, param eq=false) {
-  if eq {
-    if reverse then return a >= b;
-    else return a <= b;
+pragma "no doc"
+/* Empty record to serve as default comparator value in function arguments */
+record defaultComparator{ }
+
+pragma "no doc"
+const emptyRecord = new defaultComparator();
+
+
+pragma "no doc"
+/*
+    Compare function used for all sorts of sort functions.
+
+    Returns:
+      1 if a > b
+      0 if a == b
+     -1 if a < b
+
+ */
+private inline proc chpl_compare(a, b, param reverse=false, comparator:?rec=emptyRecord): int {
+  use Reflection;
+
+  // TODO -- we could also check types
+  // Use comparator.compare(a, b) if is defined by user
+  if canResolveMethod(comparator, "compare", a, b) {
+    if reverse { return -comparator.compare(a, b); }
+    else { return  comparator.compare(a, b); }
+  }
+
+  // TODO -- It may be faster to call key() on all data once, and mirror swaps
```

Comment:
I'm fine with this comment, but Knuth volume 3 has a good discussion. You generally will only benefit from such a transformation if the values being sorted are much larger than their keys. I actually think that's less likely than the value being close to the same size as the key.